File: todoapp/app/src/main/java/com/example/android/architecture/blueprints/todoapp/util/FragmentScoped.java
package com.example.android.architecture.blueprints.todoapp.util;

import com.example.android.architecture.blueprints.todoapp.data.source.TasksRepositoryComponent;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import javax.inject.Scope;

/**
 * In Dagger, an unscoped component cannot depend on a scoped component. As
 * {@link TasksRepositoryComponent} is a scoped component ({@code @Singleton}, we create a custom
 * scope to be used by all fragment components. Additionally, a component with a specific scope
 * cannot have a sub component with the same scope.
 */
@Documented
@Scope
@Retention(RetentionPolicy.RUNTIME)
public @interface FragmentScoped {
}

File: app/src/main/java/com/example/android/architecture/blueprints/todoapp/di/FragmentScoped.kt
package com.example.android.architecture.blueprints.todoapp.di


import javax.inject.Scope
import kotlin.annotation.AnnotationRetention.RUNTIME

@Scope
@Retention(RUNTIME)
@Target(
    AnnotationTarget.CLASS, AnnotationTarget.FILE, AnnotationTarget.FUNCTION,
    AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER
)
annotation class FragmentScoped